File: C++/Problems/Problems/problem_1358.cpp
#include "problem_1358.h"

bool problem_1358::test() {
    auto s = "abcabc";
    auto expected = 10;

    auto result = numberOfSubstrings(s);

    return expected == result;
}

int problem_1358::numberOfSubstrings(std::string s) {
    auto left = 0;
    int counts[3] = {0};
    auto result = 0;
    for (auto right = 0; right < s.size(); right++)
    {
        counts[s[right] - 'a']++;

        while (counts[0] > 0 && counts[1] > 0 && counts[2] > 0)
        {
            result += s.size() - right;
            counts[s[left] - 'a']--;
            left++;
        }
    }

    return result;
}

File: Python/Problems/Problems/problem1358.py
from .problem import Problem


class Problem1358(Problem):
    def test(self):
        s = "abcabc"
        expected = 10

        result = self.numberOfSubstrings(s)

        return expected == result

    def numberOfSubstrings(self, s):
        left = 0
        counts = [0, 0, 0]
        result = 0
        for right in range(len(s)):
            counts[ord(s[right]) - ord('a')] += 1

            while counts[0] > 0 and counts[1] > 0 and counts[2] > 0:
                result += len(s) - right
                counts[ord(s[left]) - ord('a')] -= 1
                left += 1

        return result